File: VB/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web.Mvc
Imports System.Web.Script.Serialization
Imports GridViewOptimisticConcurrencyMvc.Models

Namespace GridViewOptimisticConcurrencyMvc.Controllers
    Public Class HomeController
        Inherits Controller
        Private db As New CustomerDbContext()

        Public Function Index() As ActionResult
            Return View()
        End Function

        Public Function GridViewPartial() As ActionResult
            Return PartialView(db.Customers.ToList())
        End Function

        Public Function GridViewPartialAddNew(ByVal customer As Customer) As ActionResult
            Dim model = db.Customers

            If ModelState.IsValid Then
                Try
                    db.Entry(customer).State = System.Data.Entity.EntityState.Added
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("GridViewPartial", db.Customers.ToList())
        End Function

        Public Function GridViewPartialUpdate(ByVal customer As Customer) As ActionResult
            Dim model = db.Customers

            customer.RowVersion = CalculateOldRowVersion(customer.Id)

            If ModelState.IsValid Then
                Try
                    db.Entry(customer).State = System.Data.Entity.EntityState.Modified
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("GridViewPartial", db.Customers.ToList())
        End Function

        Public Function GridViewPartialDelete(ByVal customer As Customer) As ActionResult
            Dim model = db.Customers

            customer.RowVersion = CalculateOldRowVersion(customer.Id)

            If ModelState.IsValid Then
                Try
                    db.Entry(customer).State = System.Data.Entity.EntityState.Deleted
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("GridViewPartial", db.Customers.ToList())
        End Function

        Private Function CalculateOldRowVersion(ByVal id As Integer) As Byte()
            Dim serializer As New JavaScriptSerializer()
            Dim rowVersions As String = Request("RowVersions")
            Dim dictionary As Dictionary(Of Object, String) = CType(serializer.Deserialize(rowVersions, GetType(Dictionary(Of Object, String))), Dictionary(Of Object, String))
            Dim rowVersion() As Char = dictionary(id.ToString()).ToCharArray()

            Return Convert.FromBase64CharArray(rowVersion, 0, rowVersion.Length)
        End Function
    End Class
End Namespace

File: CS/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using System.Web.Script.Serialization;
using GridViewOptimisticConcurrencyMvc.Models;

namespace GridViewOptimisticConcurrencyMvc.Controllers {
    public class HomeController : Controller {
        private CustomerDbContext db = new CustomerDbContext();

        public ActionResult Index() {
            return View();
        }

        public ActionResult GridViewPartial() {
            return PartialView(db.Customers.ToList());
        }

        public ActionResult GridViewPartialAddNew(Customer customer) {
            var model = db.Customers;

            if (ModelState.IsValid) {
                try {
                    db.Entry(customer).State = System.Data.Entity.EntityState.Added;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("GridViewPartial", db.Customers.ToList());
        }

        public ActionResult GridViewPartialUpdate(Customer customer) {
            var model = db.Customers;

            customer.RowVersion = CalculateOldRowVersion(customer.Id);

            if (ModelState.IsValid) {
                try {
                    db.Entry(customer).State = System.Data.Entity.EntityState.Modified;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("GridViewPartial", db.Customers.ToList());
        }

        public ActionResult GridViewPartialDelete(Customer customer) {
            var model = db.Customers;

            customer.RowVersion = CalculateOldRowVersion(customer.Id);

            if (ModelState.IsValid) {
                try {
                    db.Entry(customer).State = System.Data.Entity.EntityState.Deleted;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("GridViewPartial", db.Customers.ToList());
        }

        private byte[] CalculateOldRowVersion(int id) {
            JavaScriptSerializer serializer = new JavaScriptSerializer();
            string rowVersions = Request["RowVersions"];
            Dictionary<object, string> dictionary = (Dictionary<object, string>)serializer.Deserialize(rowVersions, typeof(Dictionary<object, string>));
            char[] rowVersion = dictionary[id.ToString()].ToCharArray();

            return Convert.FromBase64CharArray(rowVersion, 0, rowVersion.Length);
        }
    }
}